polkadot/node/core/backing/src/lib.rs

```diff
@@ -981,6 +996,81 @@ async fn handle_active_leaves_update<Context>(
 	Ok(())
 }
 
+macro_rules! try_runtime_api {
+	($x: expr) => {
+		match $x {
+			Ok(x) => x,
+			Err(err) => {
+				// Only bubble up fatal errors.
+				error::log_error(Err(Into::<runtime::Error>::into(err).into()))?;
+
+				// We can't do candidate validation work if we don't have the
+				// requisite runtime API data. But these errors should not take
+				// down the node.
+				return Ok(None)
+			},
+		}
+	};
+}
+
+fn core_index_from_statement(
+	validator_groups: &[Vec<ValidatorIndex>],
+	group_rotation_info: &GroupRotationInfo,
+	cores: &[CoreState],
+	statement: &SignedFullStatementWithPVD,
+) -> Option<CoreIndex> {
+	let compact_statement = statement.as_unchecked();
+	let candidate_hash = CandidateHash(*compact_statement.unchecked_payload().candidate_hash());
+
+	let n_cores = cores.len();
+
+	gum::trace!(target: LOG_TARGET, ?group_rotation_info, ?statement, ?validator_groups, n_cores = ?cores.len() , ?candidate_hash, "Extracting core index from statement");
+
+	let statement_validator_index = statement.validator_index();
+	for (group_index, group) in validator_groups.iter().enumerate() {
```

Comment:
Given that we are doing this for each and every statement, a table ValidatorIndex -> GroupIndex might make sense (also valid for an entire session), so we get O(1) complexity here.

Reply:
done, PTAL